cmd/cstorvolumeclaim/controller.go

```diff
@@ -0,0 +1,404 @@
+	// name already existing.
+	ErrResourceExists = "ErrResourceExists"
+
+	// MessageResourceExists is the message used for Events when a resource
+	// fails to sync due to a cstorvolumeclaim already existing
+	MessageResourceExists = "Resource %q already exists and is not managed by CVC"
+	// MessageResourceSynced is the message used for an Event fired when a
+	// cstorvolumeclaim is synced successfully
+	MessageResourceSynced = "cstorvolumeclaim synced successfully"
+
+	// CStorVolumeClaimFinalizer name of finalizer on CStorVolumeClaim that
+	// are bound by CStorVolume
+	CStorVolumeClaimFinalizer = "cvc.openebs.io/finalizer"
+)
+
+// Patch struct represent the struct used to patch
+// the cstorvolumeclaim object
+type Patch struct {
+	Op    string `json:"op"`
+	Path  string `json:"path"`
+	Value string `json:"value"`
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the spcPoolUpdated resource
+// with the current status of the resource.
+func (c *CVCController) syncHandler(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing cstorvolumeclaim %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing cstorvolumeclaim %q (%v)", key, time.Since(startTime))
+	}()
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the cvc resource with this namespace/name
+	cvc, err := c.cvcLister.CStorVolumeClaims(namespace).Get(name)
+	if k8serror.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("cstorvolumeclaim '%s' has been deleted", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	cvcCopy := cvc.DeepCopy()
+	err = c.syncCVC(cvcCopy)
+	return err
+}
+
+// enqueueCVC takes a CVC resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than CStorVolumeClaims.
+func (c *CVCController) enqueueCVC(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+
+	/*	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+			obj = unknown.Obj
+		}
+		if cvc, ok := obj.(*apis.CStorVolumeClaim); ok {
+			objName, err := cache.DeletionHandlingMetaNamespaceKeyFunc(cvc)
+			if err != nil {
+				glog.Errorf("failed to get key from object: %v, %v", err, cvc)
+				return
+			}
+			glog.V(5).Infof("enqueued %q for sync", objName)
+			c.workqueue.Add(objName)
+		}
+	*/
+}
+
+// synCVC is the function which tries to converge to a desired state for the
+// CStorVolumeClaims
+func (c *CVCController) syncCVC(cvc *apis.CStorVolumeClaim) error {
```

Comment:
May be later - a little info on what is the desired state will be good to add. what kind of reconciliation will happen?

Reply:
Sure .. will do